### PR TITLE
Refactor registering flags to separate AddEmbedFlags function for future reuse

### PR DESCRIPTION
**Pull request by serathius**

### CODE REVIEW

File: server/etcdmain/config.go [hunk 1]
```diff
@@ -97,17 +97,28 @@ func newConfig() *config {
...
+
+	fs.BoolVar(&cfg.printVersion, "version", false, "Print the version and exit.")
+	// ignored
+	for _, f := range cfg.ignored {
+		fs.Var(&flags.IgnoredFlag{Name: f}, f, "")
+	}
+	return cfg
+}
 
+func AddEmbedFlags(cfg *embed.Config, fs *flag.FlagSet) {
```
Comment (ahrtr) [inline server/etcdmain/config.go:118]: Add a method `(*embed.Config) AddFlags(fs *flag.FlagSet)` under the `server/embed `package, so as to encapsulate the details in the `embed` package?

Comment (serathius) [reply server/etcdmain/config.go:118]: Right, I decided to first separate this step to make the review simpler. 
I can add a separate commit with your suggestion.

**Review by ahrtr — APPROVED** [review]

File: server/etcdmain/config.go (hunk 1 above)
Comment (serathius) [reply server/etcdmain/config.go]: Done

**Comment by serathius** [comment]
Added the change you requested. PTAL @ahrtr

**Comment by ahrtr** [comment]
In the second commit, you added about 363 lines of code in embed/config.go, but only removed about 185 lines of code in etcdmain/config.go. They should be roughly equal.

File: server/etcdmain/config.go [hunk 2]
```diff
@@ -97,200 +95,17 @@ func newConfig() *config {
...
-
-	fs.StringVar(&cfg.ec.Dproxy, "discovery-proxy", cfg.ec.Dproxy, "HTTP proxy to use for traffic to discovery service. Will be deprecated in v3.7, and be decommissioned in v3.8.")
-	fs.StringVar(&cfg.ec.DNSCluster, "discovery-srv", cfg.ec.DNSCluster, "DNS domain used to bootstrap initial cluster.")
-	fs.StringVar(&cfg.ec.DNSClusterServiceName, "discovery-srv-name", cfg.ec.DNSClusterServiceName, "Service name to query when using DNS discovery.")
-	fs.StringVar(&cfg.ec.InitialCluster, "initial-cluster", cfg.ec.InitialCluster, "Initial cluster configuration for bootstrapping.")
-	fs.StringVar(&cfg.ec.InitialClusterToken, "initial-cluster-token", cfg.ec.InitialClusterToken, "Initial cluster token for the etcd cluster during bootstrap.")
 	fs.Var(cfg.cf.clusterState, "initial-cluster-state", "Initial cluster state ('new' when bootstrapping a new cluster or 'existing' when adding new members to an existing cluster). After successful initialization (bootstrapping or adding), flag is ignored on restarts.")
```
Comment (fuweid) [inline server/etcdmain/config.go:105]: would you mind sharing the reason why we don't include this in the `// clustering`?

**Review by fuweid — APPROVED** [review]
LGTM

Using the following patch to compare main and this pr. There is no diff

```diff
diff --git a/server/etcdmain/config.go b/server/etcdmain/config.go
index 5327ff953..fc05be811 100644
--- a/server/etcdmain/config.go
+++ b/server/etcdmain/config.go
@@ -96,9 +96,6 @@ func newConfig() *config {
                        string(cconfig.V2_DEPR_2_GONE)),
        }
        fs := cfg.cf.flagSet
-       fs.Usage = func() {
-               fmt.Fprintln(os.Stderr, usageline)
-       }
```

File: server/embed/config.go [hunk 3]
```diff
@@ -553,6 +555,367 @@ func NewConfig() *Config {
...
+	fs.UintVar(&cfg.ExperimentalBootstrapDefragThresholdMegabytes, "experimental-bootstrap-defrag-threshold-megabytes", 0, "Enable the defrag during etcd server bootstrap on condition that it will free at least the provided threshold of disk space. Needs to be set to non-zero value to take effect.")
+	fs.IntVar(&cfg.ExperimentalMaxLearners, "experimental-max-learners", membership.DefaultMaxLearners, "Sets the maximum number of learners that can be available in the cluster membership.")
+	fs.DurationVar(&cfg.ExperimentalWaitClusterReadyTimeout, "experimental-wait-cluster-ready-timeout", cfg.ExperimentalWaitClusterReadyTimeout, "Maximum duration to wait for the cluster to be ready.")
+	fs.Uint64Var(&cfg.SnapshotCatchUpEntries, "experimental-snapshot-catchup-entries", cfg.SnapshotCatchUpEntries, "Number of entries for a slow follower to catch up after compacting the raft storage entries.")
+
+	// unsafe
+	fs.BoolVar(&cfg.UnsafeNoFsync, "unsafe-no-fsync", false, "Disables fsync, unsafe, will cause data loss.")
+	fs.BoolVar(&cfg.ForceNewCluster, "force-new-cluster", false, "Force to create a new one member cluster.")
+}
+func AddEmbedFlags(cfg *Config, fs *flag.FlagSet) {
```
Comment (ahrtr) [inline server/embed/config.go:738]: Remove `AddEmbedFlags`?

File: server/etcdmain/config.go (hunk 2 above)
Comment (serathius) [reply server/etcdmain/config.go:105]: They are based on `cfg.cf` and not `cfg.ec`.

File: server/embed/config.go (hunk 3 above)
Comment (serathius) [reply server/embed/config.go]: Ups, must have pasted twice and not notice.

**Comment by serathius** [comment]
Fixed

**Review by ahrtr — APPROVED** [review]

**Review by fuweid — APPROVED** [review]
LGTM

**Pull request merged**